### PR TITLE
Rewrite nose tests to use unittest

### PR DESCRIPTION
**Pull request by ievacerny**
Addresses #613 

All unit test functions have been added to `unittest.TestCase`s. Nose and bare asserts have been replace with unittest assertion methods. All tests are now compatible to be run with unittest test runner.

### CODE REVIEW

**Comment by ievacerny** [comment]
pyqt5 Travis check on Mac failed with segmentation fault. I think it just needs a restart but I don't have the permission for that.

**Comment by mdickinson** [comment]
> I think it just needs a restart but I don't have the permission for that.

Closing and re-opening the PR is often the easiest way to re-trigger the jobs.

File: traitsui/tests/ui_editors/test_data_frame_editor.py [hunk 1]
```diff
@@ -6,16 +6,15 @@
...
 from numpy.testing import assert_array_equal
 
 
 try:
     from pandas import DataFrame
-except ImportError as exc:
-    print("Can't import Pandas: skipping")
-    raise nose.SkipTest
+except ImportError:
+    raise unittest.SkipTest("Can't import Pandas: skipping")
```
Comment (kitchoi) [inline traitsui/tests/ui_editors/test_data_frame_editor.py:17]: Request change: Same as above, this prevents the module from being run without `discover`. Could we add something like `skip_if_no_pandas` or `requires_pandas` decorator in the testing tools module instead?

File: traitsui/tests/ui_editors/test_data_frame_editor.py [hunk 2]
```diff
@@ -88,411 +87,409 @@ def sample_text_data():
...
 
 
-@skip_if_null
-def test_adapter_get_item():
-    viewer = sample_data()
-    adapter = DataFrameAdapter()
+class TestDataFrameEditor(unittest.TestCase):
 
-    item_0_df = adapter.get_item(viewer, "data", 0)
+    @skip_if_null
```
Comment (kitchoi) [inline traitsui/tests/ui_editors/test_data_frame_editor.py:92]: Not a request change, just acknowledging an existing issue:
Looks like the GUI event loop is not required for this test, I am wondering if this skip is necessary at all.
That said, this is orthogonal to the purpose of this PR so it would also make sense for this to be changed separately.

Comment (corranwebster) [reply traitsui/tests/ui_editors/test_data_frame_editor.py:92]: My suspicion is that because the `DataFrameEditor` is really a TraitsUI `View` wrapped around a `TabularEditor` and inserted into a special editor type, that it requires more TraitsUI machinery than is available in the `Null` editor - it's not a question of event loops, its that the null backend doesn't implement things (not even dummy versions).

File: traitsui/tests/test_actions.py [hunk 3]
```diff
@@ -27,11 +27,16 @@
...
+    is_mac_os,
+    skip_if_not_qt4,
+    skip_if_not_wx,
+    store_exceptions_on_all_threads,
+)
 
-if _is_current_backend("null"):
-    raise nose.SkipTest("Not supported using the null backend")
+if is_current_backend_null():
+    raise unittest.SkipTest("Not supported using the null backend")
```
[the remark this answers is not in the document]
Comment (ievacerny) [reply traitsui/tests/test_actions.py:39]: This block is needed because `Action` class that is being instantiated just below is a toolkit object. Without this skip it would throw a `NotImplementedError`.

Also, I'm not sure if this PR would be the right place to address this anyways. The goal of this PR was to rewrite tests to use unittest without modifying their behaviour. So just like it wasn't possible to run this module directly with nose, it isn't possible to run it directly with unittest, even though the output is different.

[the remark this answers is not in the document]
Comment (kitchoi) [reply traitsui/tests/test_actions.py:39]: Ah I see.  I did not notice the class below can't even be instantiated without a backend.  While that's not a desirable interface, I agree that would be beyond the scope of this PR to change that.

I was hoping that I could do `python -m unittest traitsui.tests.test_actions` to see all the tests being skipped, rather than being greeted with an error. But that is very minor, and the outcome is the same: No tests cannot be run here.

File: traitsui/tests/ui_editors/test_data_frame_editor.py (hunk 1 above)
Comment (kitchoi) [reply traitsui/tests/ui_editors/test_data_frame_editor.py:17]: Unlike the instance above, I was able to run the module and see all the tests skipped if I did this:
```

try:
    from pandas import DataFrame
except ImportError:
    _HAS_PANDAS = False
else:
    _HAS_PANDAS = True


@unittest.skipIf(not _HAS_PANDAS, "pandas DataFrame is not available.")
class TestDataFrameEditor(unittest.TestCase):
...

```

```
$ python -m unittest -v traitsui.tests.ui_editors.test_data_frame_editor
test_adapter_delete_end (traitsui.tests.ui_editors.test_data_frame_editor.TestDataFrameEditor) ... skipped 'pandas DataFrame is not available.'
test_adapter_delete_end_numerical_index (traitsui.tests.ui_editors.test_data_frame_editor.TestDataFrameEditor) ... skipped 'pandas DataFrame is not available.'
...
```

Comment (kitchoi) [reply traitsui/tests/ui_editors/test_data_frame_editor.py:17]: Not raising is consistent with the behaviour of `nose`:
On master:
```
$ python -m nose.core -v traitsui.tests.ui_editors.test_data_frame_editor
/Users/kchoi/.edm/envs/traitsui-test-3.6-null/lib/python3.6/runpy.py:125: RuntimeWarning: 'nose.core' found in sys.modules after import of package 'nose', but prior to execution of 'nose.core'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
Failure: SkipTest () ... SKIP

----------------------------------------------------------------------
Ran 1 test in 0.000s

OK (SKIP=1)
```

Hence I think this is within scope to bring the behaviour aligned with the existing one. That said, I agree this only brings a slight annoyance for the developers, and it is also not hard to change.

Comment (ievacerny) [reply traitsui/tests/ui_editors/test_data_frame_editor.py:17]: I don't think it's equivalent because nose shows the whole module as one skipped test while unittest shows individual skipped tests. And total number of tests would also be different when running full test suites. So I would be inclined to leave it as is for now, and if it does cause issues, it can be changed later.

File: traitsui/tests/ui_editors/test_data_frame_editor.py (hunk 2 above)
Comment (kitchoi) [reply traitsui/tests/ui_editors/test_data_frame_editor.py:92]: On master, if I remove the `skip_if_null` for tests that don't call `edit_traits`, I get this:
```
$ python -m nose.core -v traitsui.tests.ui_editors.test_data_frame_editor
/Users/kchoi/.edm/envs/traitsui-test-3.6-null/lib/python3.6/runpy.py:125: RuntimeWarning: 'nose.core' found in sys.modules after import of package 'nose', but prior to execution of 'nose.core'; this may result in unpredictable behaviour
  warn(RuntimeWarning(msg))
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_get_item ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_empty_dataframe ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_no_rows ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_get_item_numerical ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor_alternate_adapter ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor_columns ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor_font_mapping ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor_format_mapping ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor_multi_select ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor_numerical_index ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor_text_data ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor_with_refresh ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_data_frame_editor_with_update_refresh ... SKIP: Test not working on the 'null' backend
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_delete_start ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_delete_start_numerical_index ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_delete_middle ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_delete_middle_numerical_index ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_delete_end ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_delete_end_numerical_index ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_insert_start ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_insert_start_numerical_index ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_insert_middle ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_insert_middle_numerical_index ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_insert_end ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_insert_end_numerical_index ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_set_text ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_set_text_invalid ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_set_index_text ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_set_index_text_numeric ... ok
traitsui.tests.ui_editors.test_data_frame_editor.test_adapter_set_index_text_numeric_invalid ... ok
```

I will open an issue to remove these redundant skips. It is orthogonal to this PR.

Comment (kitchoi) [reply traitsui/tests/ui_editors/test_data_frame_editor.py:92]: https://github.com/enthought/traitsui/issues/816

File: traitsui/tests/ui_editors/test_data_frame_editor.py (hunk 1 above)
Comment (kitchoi) [reply traitsui/tests/ui_editors/test_data_frame_editor.py:17]: OK, for later.

File: traitsui/tests/editors/test_range_editor_spinner.py [hunk 4]
```diff
@@ -47,104 +53,108 @@ class NumberWithSpinnerEditor(HasTraits):
...
+    def test_wx_spin_control_editing_does_not_update(self):
+        # Bug: when editing the text part of a spin control box, pressing
+        # the OK button does not update the value of the HasTraits class
+        # on Mac OS X
+
+        # But under wx >= 3.0 this has been resolved
+        import wx
+
+        if wx.VERSION >= (3, 0):
+            return
```
[the remark this answers is not in the document]
Comment (ievacerny) [reply traitsui/tests/editors/test_range_editor_spinner.py:101]: I think this also goes with the other two comments about changing the behaviour, so I think it should stay as is in this PR.

File: traitsui/tests/test_labels.py [hunk 5]
```diff
@@ -118,124 +133,112 @@ class EnableWhenDialog(HasTraits):
...
+            h_space = text_label.x() - text.x()
+            self.assertLess(h_space, 100)
+            # and the text item size should be large
+            self.assertGreater(text.width(), _DIALOG_WIDTH - 200)
 
-@skip_if_not_qt4
-def test_qt_labels_right_resizing_vertical():
-    _test_qt_labels_right_resizing(VResizeTestDialog)
+            # the size of the window should still be 500
+            self.assertEqual(ui.control.width(), _DIALOG_WIDTH)
```
[the remark this answers is not in the document]
Comment (ievacerny) [reply traitsui/tests/test_labels.py:199]: Opened https://github.com/enthought/traitsui/issues/817

File: traitsui/tests/editors/test_range_editor_spinner.py [hunk 6]
```diff
@@ -47,104 +53,108 @@ class NumberWithSpinnerEditor(HasTraits):
...
 
             # press the OK button and close the dialog
             press_ok_button(ui)
-    except AttributeError:
-        # if all went well, we should not be here
-        assert False, "AttributeError raised"
 
+            # if all went well, the number traits has been updated and its
+            # value is 4
+            print(num.number)
```
[the remark this answers is not in the document]
Comment (ievacerny) [reply traitsui/tests/editors/test_range_editor_spinner.py]: There weren't that many print statements in unit tests, so I did remove them in this PR.

**Comment by kitchoi** [comment]
https://github.com/enthought/traitsui/pull/815 included PyQt5 to the CI run with Linux on Travis CI.  Previously PyQt5 with Linux was in the allow-to-fail job, and we never ran the test with PyQt5 there because the command with wx failed first.

While the test suite passed then, this PR might have uncovered test interactions that were not revealed before.

Two options:
(1) Bump PyQt5 back to allow-failure job for Travis Linux, equivalent to not running the test suite against it because of wx.
(2) Skip that single failing test and open an issue as it would not be reasonable to fix all these in this one PR, and this PR alone is valueable in itself.

Since there is only one test failure, I am included towards option 2.

**Comment by corranwebster** [comment]
The PyQt5 tests are not in as bad a state as I had feared.  The test matrix could be set up to run PyQt5 as a separate set of tests on linux:
```
matrix:
  include:
    - env: RUNTIME=3.6 TOOLKITS="null pyside2"
    - env: RUNTIME=3.6 TOOLKITS="pyqt"
    - env: RUNTIME=3.6 TOOLKITS="wx"
    - os: osx
      env: RUNTIME=3.6 TOOLKITS="null pyqt pyside2"
    - os: osx
      env: RUNTIME=3.6 TOOLKITS="wx"
  allow_failures:
    - env: RUNTIME=3.6 TOOLKITS="wx"
    - env: RUNTIME=3.6 TOOLKITS="pyqt"
    - os: osx
      env: RUNTIME=3.6 TOOLKITS="wx"
```
should work, and would potentially give a way to work on a fix.

**Comment by ievacerny** [comment]
Went with option 1 and opened https://github.com/enthought/traitsui/pull/818 to modify Travis setup
